### PR TITLE
🌱 Use release markers in e2e config to fetch lastest releases

### PR DESCRIPTION
**Pull request by zhanggbj**
**What this PR does / why we need it**:
Use release markers in e2e config to fetch the lastest releases. Please find more details in #2549 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #2549

### CODE REVIEW

**Comment by zhanggbj** [comment]
The release markers are working well, such as in the previous run, it gets the [latest patch release of CAPI](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/2895/pull-cluster-api-provider-vsphere-e2e-govmomi-blocking-main/1777607558396121088/artifacts/repository/cluster-api/):
- v1.5.7
- v1.6.3
- v1.7.0-rc.0

However, pull-cluster-api-provider-vsphere-e2e-govmomi-blocking-main failed  with an error as below, seems related to the ClusterClass Variable matadata from cluster-api 1.7.x.  CC @sbueringer @chrischdi  Am I missing something?
```
 Error from server (BadRequest): error when creating "STDIN": ClusterClass in version "v1beta1" cannot be handled as a ClusterClass: strict decoding error: unknown field "spec.variables[0].metadata", unknown field "spec.variables[1].metadata", unknown field "spec.variables[2].metadata", unknown field "spec.variables[3].metadata", unknown field "spec.variables[4].metadata", unknown field "spec.variables[5].metadata"
```

**Comment by zhanggbj** [comment]
For pull-cluster-api-provider-vsphere-e2e-supervisor-blocking-main failure, need to check with vm-operator team, tag `v1.8.6` existed but there's no yaml file. The last success pulled yaml is `v1.8.1`

```
msg: "failed to get https://storage.googleapis.com/artifacts.k8s-staging-capi-vsphere.appspot.com/vm-operator/v1.8.6.yaml: 
```

**Comment by chrischdi** [comment]
According to the logs it did not initialise using v1.7.0-rc, but with v1.6.3 instead:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/2895/pull-cluster-api-provider-vsphere-e2e-govmomi-blocking-main/1777997015041445888/artifacts/clusters/bootstrap/clusterctl-init.log

File: test/e2e/config/vsphere.yaml [hunk 1]
```diff
@@ -34,29 +34,29 @@ providers:
   - name: cluster-api
     type: CoreProvider
     versions:
-      - name: v1.7.0
+      - name: "{go://sigs.k8s.io/cluster-api@latest-v1.7}" # latest published release in the v1beta1 series
```
Comment (chrischdi) [inline test/e2e/config/vsphere.yaml:37]: This here changes to `v1.7.0-rc.0` (or `rc.1`), which is why the init falls back to v1.6.3 for initialisation.

Comment (chrischdi) [reply test/e2e/config/vsphere.yaml:37]:
```suggestion
      - name: "v1.7.0" # latest published release in the v1beta1 series
```

Keeping these name fields to v1.7.0 as long as there are pre-releases should help. But we should find a better way IMHO 🤔

**Review by chrischdi — COMMENTED** [review]
Sorry for the late reply. I did not take a look, because the PR is WIP

File: test/e2e/config/vsphere.yaml (hunk 1 above)
Comment (zhanggbj) [reply test/e2e/config/vsphere.yaml:37]: Thanks @chrischdi for the help! Yes, it should be pulling v1.7.0-rc.x with this PR, but I didn't quite get why it falls back to v1.6.3 and what do you mean about pre-releases?

Comment (chrischdi) [reply test/e2e/config/vsphere.yaml]: It uses the latest non-pre-release version in the configuration.

Faking it to v1.7.0 does the trick here, although not being optimal.

Comment (zhanggbj) [reply test/e2e/config/vsphere.yaml]: Got it, thanks! So for this PR, I'll leave it as hard-coded.

**Comment by zhanggbj** [comment]
More details for code review:
- Latest CAPI release is still using hard-coded format as discussed above. 
- vm-operator is a fake release and maintained by capv, so I didn't change it to release markers.

CC @chrischdi @sbueringer

File: test/e2e/config/vsphere.yaml [hunk 2]
```diff
@@ -68,29 +68,29 @@ providers:
   - name: kubeadm
     type: BootstrapProvider
     versions:
-      - name: v1.7.0
+      - name: "{go://sigs.k8s.io/cluster-api@latest-v1.7}" # latest published release in the v1beta1 series
```
Comment (chrischdi) [inline test/e2e/config/vsphere.yaml:71]:
```suggestion
      - name: v1.7.0 # latest published non-pre-release in the v1beta1 series
```

File: test/e2e/config/vsphere.yaml [hunk 3]
```diff
@@ -102,29 +102,29 @@ providers:
   - name: kubeadm
     type: ControlPlaneProvider
     versions:
-      - name: v1.7.0
+      - name: "{go://sigs.k8s.io/cluster-api@latest-v1.7}" # latest published release in the v1beta1 series
```
Comment (chrischdi) [inline test/e2e/config/vsphere.yaml:105]:
```suggestion
      - name: v1.7.0 # latest published non-pre-release in the v1beta1 series
```

File: test/e2e/config/vsphere.yaml [hunk 4]
```diff
@@ -34,7 +34,7 @@ providers:
   - name: cluster-api
     type: CoreProvider
     versions:
-      - name: v1.7.0
+      - name: v1.7.0 # latest published non-pre-release in the v1beta1 series
         # Use manifest from source files
         value: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.7.0-rc.0/core-components.yaml"
```
Comment (chrischdi) [inline test/e2e/config/vsphere.yaml:39]:
```suggestion
        value: "https://github.com/kubernetes-sigs/cluster-api/releases/download/{go://sigs.k8s.io/cluster-api@latest-v1.7}/core-components.yaml"
```

**Review by chrischdi — COMMENTED** [review]
According logs it would install mixed capi versions otherwise!

```
...
Installing Provider="cluster-api" Version="v1.7.0" TargetNamespace="capi-system"
Creating objects Provider="cluster-api" Version="v1.7.0" TargetNamespace="capi-system"
...
Installing Provider="bootstrap-kubeadm" Version="v1.6.3" TargetNamespace="capi-kubeadm-bootstrap-system"
Creating objects Provider="bootstrap-kubeadm" Version="v1.6.3" TargetNamespace="capi-kubeadm-bootstrap-system"
...
Installing Provider="control-plane-kubeadm" Version="v1.6.3" TargetNamespace="capi-kubeadm-control-plane-system"
Creating objects Provider="control-plane-kubeadm" Version="v1.6.3" TargetNamespace="capi-kubeadm-control-plane-system"
...
```

source: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/2895/pull-cluster-api-provider-vsphere-e2e-govmomi-blocking-main/1779704913966665728/artifacts/clusters/bootstrap/clusterctl-init.log

File: test/e2e/config/vsphere.yaml (hunk 4 above)
Comment (zhanggbj) [reply test/e2e/config/vsphere.yaml]: Nice catch! Just fix it.

File: test/e2e/config/vsphere.yaml (hunk 2 above)
Comment (zhanggbj) [reply test/e2e/config/vsphere.yaml]: Fixed.

File: test/e2e/config/vsphere.yaml (hunk 3 above)
Comment (zhanggbj) [reply test/e2e/config/vsphere.yaml]: Fixed.

File: test/e2e/config/vsphere.yaml [hunk 5]
```diff
@@ -34,29 +34,29 @@ providers:
   - name: cluster-api
     type: CoreProvider
     versions:
-      - name: v1.7.0
+      - name: v1.7.0 # latest published non-pre-release in the v1beta1 series
         # Use manifest from source files
-        value: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.7.0-rc.0/core-components.yaml"
+        value: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.7.0/core-components.yaml"
```
Comment (chrischdi) [inline test/e2e/config/vsphere.yaml:39]: We still want to use `{go://sigs.k8s.io/cluster-api@latest-v1.7}` here I think. (also for the other v1.7.0 ones)

File: test/e2e/config/vsphere.yaml [hunk 6]
```diff
@@ -34,29 +34,29 @@ providers:
   - name: cluster-api
```
Comment (sbueringer) [inline test/e2e/config/vsphere.yaml:34]: I think we also have to make changes to clusterctl_upgrade_test.go (compare with: https://github.com/kubernetes-sigs/cluster-api/pull/9265/files)

File: test/e2e/config/vsphere.yaml (hunk 5 above)
Comment (zhanggbj) [reply test/e2e/config/vsphere.yaml]: @chrischdi Have to change back to hardcoded format as [here](https://github.com/kubernetes-sigs/cluster-api/pull/9265/files#diff-3ed7f0b5549fcb0a0881b8d4ff4d7cd843cc4506fb014fc460019645b0cba01bR270)

File: test/e2e/config/vsphere.yaml (hunk 6 above)
Comment (zhanggbj) [reply test/e2e/config/vsphere.yaml:34]: Thanks for catching it! Fixed.

**Comment by zhanggbj** [comment]
All comments are addressed, this is ready for another review :-)

**Comment by zhanggbj** [comment]
Probably missed test pull-cluster-api-provider-vsphere-e2e-govmomi-main last time.
Now all tests passed.

File: test/e2e/config/vsphere.yaml [hunk 7]
```diff
@@ -102,29 +102,29 @@ providers:
   - name: kubeadm
     type: ControlPlaneProvider
     versions:
-      - name: v1.7.0
+      - name: v1.7.0 # latest published non-pre-release in the v1beta1 series
```
Comment (chrischdi) [inline test/e2e/config/vsphere.yaml:105]: v1.7.0 is released, we now can use `{go://sigs.k8s.io/cluster-api@v1.7}` :-)

**Comment by sbueringer** [comment]
/hold

until after the CAPV v1.10 release next week
(we have to bump & test a lot of stuff until then, I don't want to throw this PR into the mix)

**Comment by chrischdi** [comment]
@zhanggbj can you rebase this :-)

I think it is time to get this in 🎉

**Comment by zhanggbj** [comment]
@chrischdi Done and all tests passed.

File: test/e2e/clusterctl_upgrade_test.go [hunk 8]
```diff
@@ -80,3 +110,9 @@ var _ = Describe("When testing clusterctl upgrades using ClusterClass (CAPV 1.8=
 		})
 	}, WithIP("WORKLOAD_CONTROL_PLANE_ENDPOINT_IP"))
 })
+
+// GetStableReleaseOfMinor returns the latest stable version of minorRelease.
+func GetStableReleaseOfMinor(ctx context.Context, releaseMarkerPrefix, minorRelease string) (string, error) {
```
Comment (chrischdi) [inline test/e2e/clusterctl_upgrade_test.go:115]:
```suggestion
func getStableReleaseOfMinor(ctx context.Context, releaseMarkerPrefix, minorRelease string) (string, error) {
```

File: test/e2e/clusterctl_upgrade_test.go [hunk 9]
```diff
@@ -17,14 +17,36 @@ limitations under the License.
...
+	capvReleaseMarkerPrefix = "go://sigs.k8s.io/cluster-api-provider-vsphere@v%s"
 )
 
 var _ = Describe("When testing clusterctl upgrades using ClusterClass (CAPV 1.9=>current, CAPI 1.6=>1.7) [ClusterClass]", func() {
 	const specName = "clusterctl-upgrade-1.9-current" // prefix (clusterctl-upgrade) copied from CAPI
 	Setup(specName, func(testSpecificSettingsGetter func() testSettings) {
 		capi_e2e.ClusterctlUpgradeSpec(ctx, func() capi_e2e.ClusterctlUpgradeSpecInput {
+			// Get CAPI v1.6 latest stable release
+			capiVersion16 := "1.6"
+			capiStableRelease, err := GetStableReleaseOfMinor(ctx, capiReleaseMarkerPrefix, capiVersion16)
```
Comment (chrischdi) [inline test/e2e/clusterctl_upgrade_test.go:44]:
```suggestion
			capiStableRelease16, err := GetStableReleaseOfMinor(ctx, capiReleaseMarkerPrefix, capiVersion16)
```

File: test/e2e/clusterctl_upgrade_test.go [hunk 10]
```diff
@@ -17,14 +17,36 @@ limitations under the License.
...
+			Expect(err).ToNot(HaveOccurred(), "Failed to get stable version for minor release : %s", capiVersion16)
+			// Get CAPV v1.9 latest stable release
+			capvVersion19 := "1.9"
+			capvStableRelease, err := GetStableReleaseOfMinor(ctx, capvReleaseMarkerPrefix, capvVersion19)
```
Comment (chrischdi) [inline test/e2e/clusterctl_upgrade_test.go:48]:
```suggestion
			capvStableRelease19, err := GetStableReleaseOfMinor(ctx, capvReleaseMarkerPrefix, capvVersion19)
```

File: test/e2e/clusterctl_upgrade_test.go (hunk 9 above)
Comment (chrischdi) [reply test/e2e/clusterctl_upgrade_test.go:44]: same below in other test case

**Review by chrischdi — COMMENTED** [review]
Some last nits

**Comment by sbueringer** [comment]
Thank you very much!!

/lgtm
/approve

/cherry-pick release-1.10

**Pull request merged**